File: maze.hs
-- Maze generator and solver in Haskell
-- Joe Wingbermuehle
-- 20070602 <> 20140815

module Main where

import Control.Monad
import Control.Monad.State
import Control.Monad.Reader
import Data.Map
import System.Random

data ElementType = Space | Wall | Marked | Visited | Door
    deriving (Eq)

data Direction = DLeft | DRight | DUp | DDown
    deriving (Enum)

data MazeData = MazeData {
    width :: Int,
    height :: Int,
    gen :: StdGen,
    maze :: Map (Int, Int) ElementType
}

instance Show ElementType where
    show Space = "  "
    show Wall = "[]"
    show Marked = "++"
    show Visited = show Space
    show Door = show Space

instance Show MazeData where
    show = showMaze 0 0
        where
            showMaze x y md =
                if x == (width md) then
                    if y + 1 == (height md) then ""
                    else "\n" ++ (showMaze 0 (y + 1) md)
                else
                    let e = findWithDefault Wall (x, y) (maze md) in
                    (show e) ++ (showMaze (x + 1) y md)

type MazeState a = ReaderT (Int, Int) (State MazeData) a

-- Initialize an uncarved maze.
initMaze :: Int -> Int -> Int -> MazeData
initMaze w h s =
    let xs = [0 .. w - 1] in
    let ys = [0 .. h - 1] in
    let top = [(x, 0) | x <- xs] in
    let bottom = [(x, h - 1) | x <- xs] in
    let left = [(0, y) | y <- ys] in
    let right = [(w - 1, y) | y <- ys] in
    let start = [(2, 2)] in
    let spaces = Prelude.foldl (++) [] [top, bottom, left, right, start] in
    let m1 = fromList [(p, Space) | p <- spaces] in
    let doors = [(2, 2), (2, 1), (w - 3, h - 2)] in
    let m2 = fromList [(p, Door) | p <- doors] in
    MazeData {
        width = w, height = h, gen = mkStdGen s,
        maze = union m1 m2
    }

-- Get the next position in the specified direction.
getPosition :: (Int, Int) -> Direction -> (Int, Int)
getPosition (x, y) DLeft = (x - 1, y)
getPosition (x, y) DRight = (x + 1, y)
getPosition (x, y) DUp = (x, y - 1)
getPosition (x, y) DDown = (x, y + 1)

-- Move in the specified direction.
move :: ElementType -> Direction -> MazeState (Int, Int)
move fill d = do
    pos <- ask; st <- get
    let mid = getPosition pos d
    let m = insert mid fill $ maze st
    let next = getPosition mid d
    put $ st { maze = insert next fill m }
    return next

-- Determine if we can carve in the specified direction.
canMove :: ElementType -> Direction -> MazeState Bool
canMove match d = do
    pos <- ask; st <- get
    let mid = getPosition pos d
    let e1 = findWithDefault Wall mid $ maze st
    let next = getPosition mid d
    let e2 = findWithDefault Wall next $ maze st
    return $ e1 == match && e2 == match

-- Carve the maze starting at the speciified position.
carve :: MazeState ()
carve = do
    st <- get
    let (start, g') = randomR (0, 3) $ gen st
    put $ st { gen = g' }
    mapM_ tryCarve [toEnum ((start + i) `mod` 4) | i <- [0 .. 3]]
    where
        tryCarve d = do
            cm <- canMove Wall d
            when cm (move Space d >>= (\n -> local (const n) carve))

-- Find a path from the current position to the end position.
solveTo :: (Int, Int) -> MazeState Bool
solveTo stop = do
    pos <- ask
    if pos == stop then
        return True
    else do
        st <- get; pos <- ask
        put $ st { maze = insert pos Marked (maze st) }
        foldM trySolve False [toEnum i | i <- [0 .. 3]]
    where
        trySolve result d
            | result = return True
            | otherwise = do
                cm <- canMove Space d
                if cm then do
                    next <- move Marked d
                    found <- local (const next) $ solveTo stop
                    if found then return True
                    else (move Visited d >> return False)
                else return False

-- Generate a random maze.
generate :: Int -> Int -> Int -> MazeData
generate w h s = execState (runReaderT carve (2, 2)) (initMaze w h s)

-- Solve a maze.
solve :: MazeData -> MazeData
solve md =
    let start = (2, 2) in
    let stop = ((width md) - 3, (height md - 3)) in
    execState (runReaderT (solveTo stop) start) md

-- Generate and display a solved random maze.
main :: IO ()
main = do
    let maze = generate 39 23 5
    let solved = solve maze
    putStrLn $ show maze
    putStrLn $ show solved

File: maze.cpp
// Maze generator in C++
// Joe Wingbermuehle 2013-11-15

#include <algorithm>
#include <cmath>
#include <iostream>
#include <vector>

class Maze {
public:

   /** Constructor. */
   Maze(unsigned width, unsigned height) :
      m_width(width),
      m_height(height)
   {
      m_maze.resize(m_width * m_height);
   }

   /** Generate a random maze. */
   void Generate()
   {
      Initialize();
      Carve(2, 2);
      m_maze[m_width + 2] = true;
      m_maze[(m_height - 2) * m_width + m_width - 3] = true;
   }

   /** Display the maze. */
   void Show()
   {
      for(unsigned y = 0; y < m_height; y++) {
         for(unsigned x = 0; x < m_width; x++) {
            std::cout << (m_maze[y * m_width + x] ? "  " : "[]");
         }
         std::cout << "\n";
      }
   }

private:

   /** Initialize the maze array. */
   void Initialize()
   {
      std::fill(m_maze.begin(), m_maze.end(), false);
      for(unsigned x = 0; x < m_width; x++) {
         m_maze[x] = true;
         m_maze[(m_height - 1) * m_width + x] = true;
      }
      for(unsigned y = 0; y < m_height; y++) {
         m_maze[y * m_width] = true;
         m_maze[y * m_width + m_width - 1] = true;
      }
   }

   /** Carve starting at x, y. */
   void Carve(int x, int y)
   {
      m_maze[y * m_width + x] = true;
      const unsigned d = std::rand();
      for(unsigned i = 0; i < 4; i++) {
         const int dirs[] = { 1, -1, 0, 0 };
         const int dx = dirs[(i + d + 0) % 4];
         const int dy = dirs[(i + d + 2) % 4];
         const int x1 = x + dx, y1 = y + dy;
         const int x2 = x1 + dx, y2 = y1 + dy;
         if(!m_maze[y1 * m_width + x1] && !m_maze[y2 * m_width + x2]) {
            m_maze[y1 * m_width + x1] = true;
            Carve(x2, y2);
         }
      }
   }

   const unsigned m_width;
   const unsigned m_height;
   std::vector<bool> m_maze;

};

/** Generate and display a random maze. */
int main(int argc, char *argv[])
{
   Maze m(39, 23);
   m.Generate();
   m.Show();
   return 0;
}
